refactor: prohibit direct flags access in CCoinsCacheEntry and remove invalid tests

l0rinc opened the pull request:
Similarly to https://github.com/bitcoin/bitcoin/pull/30849, this cleanup is intended to de-risk https://github.com/bitcoin/bitcoin/pull/30673#discussion_r1739909068 by simplifying the coin cache public interface.

`CCoinsCacheEntry` provided general access to its internal flags state, even though, in reality, it could only be `clean`, `fresh`, `dirty`, or `fresh|dirty` (in the follow-up, we will remove `fresh` without `dirty`).

Once it was marked as `dirty`, we couldn’t set the state back to clean with `AddFlags(0)`—tests explicitly checked against that.

This PR refines the public interface to make this distinction clearer and to make invalid behavior impossible, rather than just checked by tests. We don't need extensive access to the internals of `CCoinsCacheEntry`, as many tests were simply validating invalid combinations in this way.

The last few commits contain significant test refactorings to make `coins_tests` easier to change in follow-ups.

l0rinc committed:
coins, refactor: Split up AddFlags to remove invalid states

CCoinsCacheEntry provided general access to its internal flags state, even though in reality it could only be clean, fresh, dirty or fresh|dirty.

After it got dirtied we couldn't set the state back to clean by AddFlags(0) - tests were explicitly checking against that.

This commit cleans up the public interface to make this distinction cleaner and invalid behavior impossible instead of just checked by tests.
This includes the removal of redundant `inline` qualifiers (we're inside a struct).
Also renamed `self` to `pair` to simplify the upcoming commits.

Also modernized `EmplaceCoinInternalDANGER` since it was already modified.

Co-authored-by: Andrew Toth <andrewstoth@gmail.com>

l0rinc committed:
coins, refactor: Remove direct GetFlags access

We don't need so much access to the internals of CCoinsCacheEntry, since many tests are just exercising invalid combinations this way.
This implies that `AddFlags` has private access now.

andrewtoth commented on src/test/coins_tests.cpp:
```diff
@@ -616,7 +617,9 @@ void GetCoinsMapEntry(const CCoinsMap& map, CAmount& value, char& flags, const C
+        flags = 0;
+        if (it->second.IsDirty()) flags |= DIRTY;
+        if (it->second.IsFresh()) flags |= FRESH;
         assert(flags != NO_ENTRY);
```
This assertion is obviated now.

l0rinc replied:
Replaced it with a flag validation

andrewtoth commented on src/test/coins_tests.cpp:
```diff
@@ -708,7 +711,7 @@ static void CheckSpendCoins(CAmount base_value, CAmount cache_value, CAmount exp
     BOOST_CHECK_EQUAL(result_value, expected_value);
     BOOST_CHECK_EQUAL(result_flags, expected_flags);
-};
+}
```
Why was this changed?

l0rinc replied:
Because class declarations don't need semicolons, but I'll revert if you find it distracting

andrewtoth commented on src/test/coinscachepair_tests.cpp:
```diff
@@ -146,70 +146,48 @@ BOOST_AUTO_TEST_CASE(linked_list_add_flags)
     BOOST_CHECK_EQUAL(n2.second.Next(), &sentinel);
     BOOST_CHECK_EQUAL(n2.second.Prev(), &sentinel);
 
     // But adding DIRTY re-inserts it after n2
```
`But` can be removed now.

l0rinc replied:
Because class declarations don't need semicolons, but I'll revert if you find it distracting

l0rinc replied:
Done

l0rinc replied:
Was already removed in `coins: Remove direct GetFlags access`, but I've moved it up to `coins, refactor: Split up AddFlags to remove invalid states`

andrewtoth reviewed (commented):
Concept ACK

Can we remove any other tests that are trying to add flags that are not FRESH or DIRTY and so are now useless?

l0rinc committed:
test: Group values and states in tests into CoinEntry wrappers

Note: that this commit affects the test order, but doesn't change its behavior or coverage otherwise.

l0rinc commented:
> Can we remove any other tests that are trying to add flags that are not FRESH or DIRTY and so are now useless?

I've merged `NO_ENTRY` and `char(0)` values - this simplifies the cases somewhat.
Do you think we can remove any of the lines? There aren't any duplicates, even after the above change...

andrewtoth commented on src/test/coins_tests.cpp:
```diff
@@ -571,11 +571,19 @@ const static CAmount VALUE2 = 200;
 const static CAmount VALUE3 = 300;
 const static char DIRTY = CCoinsCacheEntry::DIRTY;
 const static char FRESH = CCoinsCacheEntry::FRESH;
-const static char NO_ENTRY = -1;
```
Since we don't have pass a bitmap as flags anymore, does it make sense to get rid of the `char` type for flags in `coins_tests` and replace with a more descriptive enum?
```C++
enum Flags
{
   Clean,
   Dirty,
   Fresh,
   DirtyAndFresh
};
```

l0rinc replied:
what's the difference between clean and NO_ENTRY? Does that difference still make sense after this change?

> get rid of the char type for flags in coins_tests and replace with a more descriptive enum

Since we don't have flags in the production code anymore I would prefer getting rid of them in the tests as well. Reintroducing an enum would kinda' defeat that purpose in my opinion.

To make sure the tests are still checking the same combinations, I have added a few scripted diffs to transition the tests away from the flags - now the tabular data displays dirty and fresh booleans instead (I guess we can group some of the examples into `CoinMapEntry` as well, if needed).
This should help us selectively decide which combinations are still valid.
What do you think?

andrewtoth replied:
`Clean` means not dirty and not fresh. `NO_ENTRY` is a special value that is not a possible flag but signals to the tests that that entry does not exist. With proper types, that would correspond to a `std::nullopt`.
We don't have to name the enum `Flags`, we can name it `EntryState`. That way we don't have to validate the `flags` variable to see if it's one of the values we will accept. That is better served by using an enum and std::optional.

andrewtoth replied:
I don't think replacing with true and false is the right approach. Using enums, tests would be like
```
CheckAccessCoin(VALUE1, SPENT , SPENT , DirtyAndFresh, DirtyAndFresh);
CheckSpendCoins(ABSENT, SPENT , ABSENT, Fresh        , std::nullopt );
```

l0rinc replied:
> is a special value that is not a possible flag but signals to the tests that that entry does not exist

> That is better served by using an enum and std::optional.

Thanks, I'll investigate tomorrow.
Please leave some comments about the overall direction as well, do you think this new test layout is more readable or less?

l0rinc replied:
Did a significant restructure, the tabular tests should be a lot easier to comprehend and modify now.
I still need to get rid of remaining flags in the tests, will do that tomorrow - any early feedback until then is appreciated.

l0rinc replied:
Done

l0rinc committed:
test, refactor: Migrate GetCoinsMapEntry to return MaybeCoin

l0rinc committed:
test, refactor: Remove remaining unbounded flags from coins_tests

andrewtoth commented on src/test/coins_tests.cpp:
```diff
@@ -562,6 +564,15 @@ BOOST_AUTO_TEST_CASE(ccoins_serialization)
 
+struct CoinStruct {
+    const CAmount value;
+    const char flags;
```
I still think this should be an enum named `EntryState` or something, with all combinations of flags like https://github.com/bitcoin/bitcoin/pull/30906/commits/ba52ebd41673abf63d4495215c84be7244469d30#r1760349038. That's the proper way to do this instead of defining static vars for each value of a char we care about.

l0rinc replied:
Absolutely, that's ny plan for tomorrow - but needed to clean up the test before doing that

davidgumberg commented on src/test/coins_tests.cpp:
```diff
@@ -827,75 +805,77 @@ void CheckWriteCoins(CoinStruct parent_coin_state, CoinStruct child_coin_state,
+    CheckWriteCoins({VALUE1, DIRTY      }, {VALUE2, DIRTY      }, {VALUE2, DIRTY      });
+    CheckWriteCoins({VALUE1, DIRTY      }, {VALUE2, DIRTY|FRESH}, {FAIL  , NO_ENTRY   });
+    CheckWriteCoins({VALUE1, DIRTY|FRESH}, {VALUE2, DIRTY      }, {VALUE2, DIRTY|FRESH});
+    CheckWriteCoins({VALUE1, DIRTY|FRESH}, {VALUE2, DIRTY|FRESH}, {FAIL  , NO_ENTRY   });
```
Could be scripted diff:

```bash
sed -i -E 's/CheckWriteCoins\(([^,]+), ([^,]+), ([^,]+), ([^,]+), ([^,]+), ([^,]+)\);/CheckWriteCoins({\1, \4}, {\2, \5}, {\3, \6});/' src/test/coins_tests.cpp
```

davidgumberg commented on src/test/coins_tests.cpp:
```diff
@@ -663,47 +663,43 @@ static void CheckAccessCoin(CAmount base_value, CoinStruct coin_state, CoinStruc
+    CheckAccessCoin(VALUE1, {VALUE2, UNCHANGED  }, {VALUE2, UNCHANGED  });
+    CheckAccessCoin(VALUE1, {VALUE2, FRESH      }, {VALUE2, FRESH      });
+    CheckAccessCoin(VALUE1, {VALUE2, DIRTY      }, {VALUE2, DIRTY      });
+    CheckAccessCoin(VALUE1, {VALUE2, DIRTY|FRESH}, {VALUE2, DIRTY|FRESH});
```
Nit: Could be made into a loop as below

davidgumberg commented:
Concept ACK on dropping the bitfield interface for `CCoinsCacheEntry` flags

l0rinc committed:
test: Validate error messages on fail

The `ccoins_add` and `ccoins_write` tests check the actual exception error messages now instead of just that they fail for the given parameters.
This enables us testing different exceptions in a more fine-grained way in later changes.

l0rinc committed:
test, refactor: Compact ccoins_access and ccoins_spend

Also added an extra check for `AccessCoin` in `CheckAccessCoin` to make sure its result is also validated.

andrewtoth commented on src/coins.h:
```diff
@@ -159,19 +175,13 @@ struct CCoinsCacheEntry
+    {
+        AddFlags(FRESH, self, sentinel);
     }
     inline void ClearFlags() noexcept
```
```suggestion
    inline void SetClean() noexcept
```

l0rinc replied:
Excellent, fixed.

andrewtoth replied:
*fresh* is actually a substate of *dirty*. Being *fresh* implies *dirty*, so cleaning implies no longer fresh..

l0rinc replied:
`FRESH` is definitely a [very confusing concept](https://github.com/bitcoin/bitcoin/blob/master/src/coins.h#L143-L151)
> the parent cache does not have this coin or that it is a spent coin in the parent cache

Maybe after https://github.com/bitcoin/bitcoin/pull/30673 we can find a better name (and description) for it

hodlinator replied:
Okay, so `FRESH` and `DIRTY_FRESH` will merge into one. :+1:

l0rinc replied:
rather `FRESH` will be removed in the mentioned follow-up, since it's not actually possible (unless in case of reorgs, which I don't fully understand yet, but not important for this PR)

mrsatoshinakamotoisback replied:
My name is Robert Hache  and Karen Belzile  my system was made under attack by a group of terrorist attackers and at the. Moment I m stuck as I need someone  to help me to compl the repair wirh ny team of IT engineer  witch is curnt working on my case ... I required a secondary deposit of 240 USD to complete the repair of the system that's has Ben generated with my own request as I wS made the hub central of the attack witxh disturbed the Ecosystem  security and several other people in this matter ...as I was made their hub central of the attack...

My phone number is the 2262080980

And the subject is still lock from the blockchain  explorer  at the moment  as im worried about all the other witch were affected by this group as several individuals  and businesses ess were made affected by this same group... the phi e numbet of the 2265521155 is the primary person responsible  for this attack witch is also  known as the bad satoshi . As I have all the evidence of it and hes involve in money lowering and severe attack in my intie system... I must find a resolution for the funds missing to complete the repair witch is intended to return everything  to the party affected by the same group...ians for doing so I need a contribution witch would be payed with major interest within 48 hours of the contribution. 


Please adress  any coment directly  to my phone  number

Sincerely  best regards  Rober and engineer

mrsatoshinakamotoisback replied:
My name is Robert Hache  and Karen Belzile  my system was made under attack by a group of terrorist attackers and at the. Moment I m stuck as I need someone  to help me to compl the repair wirh ny team of IT engineer  witch is curnt working on my case ... I required a secondary deposit of 240 USD to complete the repair of the system that's has Ben generated with my own request as I wS made the hub central of the attack witxh disturbed the Ecosystem  security and several other people in this matter ...as I was made their hub central of the attack...

My phone number is the 2262080980

And the subject is still lock from the blockchain  explorer  at the moment  as im worried about all the other witch were affected by this group as several individuals  and businesses ess were made affected by this same group... the phi e numbet of the 2265521155 is the primary person responsible  for this attack witch is also  known as the bad satoshi . As I have all the evidence of it and hes involve in money lowering and severe attack in my intie system... I must find a resolution for the funds missing to complete the repair witch is intended to return everything  to the party affected by the same group...ians for doing so I need a contribution witch would be payed with major interest within 48 hours of the contribution. 


Please adress  any coment directly  to my phone  number

Sincerely  best regards  Rober and engineer

l0rinc replied:
These definitions aren't fully consistent currently, the follow-up PR will partially clean this up I think (cc: @andrewtoth )

andrewtoth commented on src/test/coins_tests.cpp:
```diff
@@ -589,314 +632,245 @@ static void SetCoinsValue(CAmount value, Coin& coin)
+    if (auto it{map.find(outp)}; it != map.end()) {
+        return CoinStruct{
+            it->second.coin.IsSpent() ? SPENT : it->second.coin.out.nValue,
+            ToCaching(it->second.IsDirty(), it->second.IsFresh())};
```
Do we need a new function if we're only calling it the one time here? Just seems like unnecessary misdirection.

l0rinc replied:
Converting flags to enums isn't the responsibility of this method, that's why I've extracted it.

hodlinator reviewed (commented):
Concept ACK.

Really like the move to put restrictions on how flags are used.

l0rinc commented on src/test/coins_tests.cpp:
```diff
@@ -562,20 +564,61 @@ BOOST_AUTO_TEST_CASE(ccoins_serialization)
+
+struct CoinStruct {
+    const CAmount value;
+    const Caching caching;
```
Had a few versions before, played around with it a bit more, thanks for the pointers, what do you think of the new structure and names.

l0rinc replied:
The usages are a lot more verbose, but I've extracted everything important now, so the tables are still readable

l0rinc commented on src/coins.h:
```diff
@@ -128,6 +128,22 @@ struct CCoinsCacheEntry
+    //! flagged pair linked list.
+    inline void AddFlags(uint8_t flags, CoinsCachePair& self, CoinsCachePair& sentinel) noexcept
+    {
+        assert(flags & DIRTY || flags & FRESH);
```
Clever, thanks!

andrewtoth reviewed (commented):
I think the commit title `test: Migrate GetCoinsMapEntry to return std::optional<CoinState>` is no longer accurate.

I didn't find splitting the changes in `coins_tests` up into that many commits to be helpful for review. I ended up reviewing the combined diff for that file. Will review in more depth later.

l0rinc commented on src/coins.h:
```diff
@@ -156,24 +172,18 @@ struct CCoinsCacheEntry
 
-    //! Adding a flag also requires a self reference to the pair that contains
-    //! this entry in the CCoinsCache map and a reference to the sentinel of the
-    //! flagged pair linked list.
```
This was moved to the `AddFlags` already - even though it seems to me the code already makes all of that clear:

> Adding a flag also requires a self reference to the pair that contains this entry in the CCoinsCache map

`AddFlags(uint8_t flags, CoinsCachePair& self`

> and a reference to the sentinel of the flagged pair linked list

, `CoinsCachePair& sentinel`

andrewtoth replied:
Ah yes I see it was moved up with `AddFlags` to private. Maybe better to keep it on `SetDirty`.

> Adding a flag also requires a self reference to the pair that contains this entry in the CCoinsCache map

There's no context in the method signature that it is an entry in the `CCoinsCache` map.

> and a reference to the sentinel of the flagged pair linked list

There's no context in the method signature about a flagged linked list.

l0rinc replied:
> There's no context in the method signature that it is an entry in the CCoinsCache map

isn't that what `CoinsCachePair` means?

> method signature about a flagged linked list

The sentinel hints at it.

But, again, I kept the methods for AddFlags - and don't want to duplicate it for SetFresh and  SetDirty.

l0rinc commented on src/coins.h:
```diff
@@ -128,6 +128,22 @@ struct CCoinsCacheEntry
+    //! flagged pair linked list.
+    inline void AddFlags(uint8_t flags, CoinsCachePair& self, CoinsCachePair& sentinel) noexcept
+    {
+        assert(flags & (DIRTY | FRESH));
```
This was meant to be a temporary transition, since we're removing the rest of the flags in your PR - but I'm fine with Assume as well - changed.

andrewtoth commented on src/test/coins_tests.cpp:
```diff
@@ -562,20 +564,57 @@ BOOST_AUTO_TEST_CASE(ccoins_serialization)
+    const CAmount value;
+    const State state;
+
+    CoinEntry(CAmount v, State s) : value(v), state(s) {}
```
```suggestion
    constexpr CoinEntry(CAmount v, State s) : value(v), state(s) {}
```
then all `const static MaybeCoin` declarations below can be made `constexpr`.

l0rinc replied:
Done, thanks

andrewtoth commented on src/test/coins_tests.cpp:
```diff
@@ -589,314 +628,245 @@ static void SetCoinsValue(CAmount value, Coin& coin)
-    CheckAddCoin(VALUE2, VALUE3, FAIL  , DIRTY|FRESH, NO_ENTRY   , false);
-    CheckAddCoin(VALUE2, VALUE3, VALUE3, DIRTY|FRESH, DIRTY|FRESH, true );
+    for (auto base_value : {ABSENT, SPENT, VALUE1}) {
+        CheckAddCoin(base_value, VALUE3, MISSING,            VALUE3_DIRTY_FRESH,   false);
```
Can we just hardcode value3 for write? It's the same for every case here.

l0rinc replied:
Wouldn't that be inconsistent?
In every other case, the constant contained both values.

andrewtoth replied:
I mean, if you look at the `Write` column, each value is `VALUE3`.

l0rinc replied:
I thought of that, but since it has to be different from `base_value`, seems weird to assume it that in a different method instead of the call-site.
```
             Base        Write   Cache               Expected              Coinbase
CheckAddCoin(base_value, VALUE3, SPENT_DIRTY,        VALUE3_DIRTY,         false); // VALUE3 + VALUE3_DIRTY -> VALUE3_DIRTY
```
makes more sense than
```
             Base        Cache               Expected              Coinbase
CheckAddCoin(base_value, SPENT_DIRTY,        VALUE3_DIRTY,         false); // SPENT_DIRTY -> VALUE3_DIRTY???
```

But I don't mind renaming `VALUE3` to `VAL3` if you're worried about verbosity.

andrewtoth commented on src/test/coins_tests.cpp:
```diff
@@ -562,20 +564,57 @@ BOOST_AUTO_TEST_CASE(ccoins_serialization)
+constexpr MaybeCoin VALUE3_DIRTY       = CoinEntry{VALUE3, CoinEntry::State::DIRTY};
+constexpr MaybeCoin VALUE3_DIRTY_FRESH = CoinEntry{VALUE3, CoinEntry::State::DIRTY_FRESH};
+
+const static std::string EX_OVERWRITE_UNSPENT = "Attempted to overwrite an unspent coin (when possible_overwrite is false)";
```
Since #30921, we can now make these `constexpr string_view`.

l0rinc replied:
Made it a `constexpr auto` (I don't want non-owning fields) https://github.com/bitcoin/bitcoin/compare/1dd79a32c9c6e0c6cf4ecef00b56972574babced..cb9be6729c4e27bf2d6d703c0d454a22cbcdb6e1

andrewtoth commented:
crACK cb9be6729c4e27bf2d6d703c0d454a22cbcdb6e1

This simplifies the CCoinsCacheEntry interface, which was leaking the flags bitfield as an implementation detail. Now we don't have to worry about fuzzing that, and it will make removing the FRESH-but-not-DIRTY or FRESH-and-spent cases easier in https://github.com/bitcoin/bitcoin/pull/30673.

Also, the tests are much saner now and easier to parse by also removing the flags interface there.

There are some minor issues with the commits though.

Commit titles `test: Migrate GetCoinsMapEntry to return std::optional<CoinState>`
and `test: Group values and states in tests into CoinState wrappers` should reference `CoinEntry` now.

`test: Validate error messages on fail` should add some context in the commit message or description that it is for CCoinsViewCache AddCoin and WriteCoin unit tests.

l0rinc commented:
No code changes since https://github.com/bitcoin/bitcoin/commit/cb9be6729c4e27bf2d6d703c0d454a22cbcdb6e1, only commit messages (as requested) and rebased on latest master (separately).

laanwj commented on src/coins.cpp:
```diff
@@ -107,13 +108,8 @@ void CCoinsViewCache::AddCoin(const COutPoint &outpoint, Coin&& coin, bool possi
 void CCoinsViewCache::EmplaceCoinInternalDANGER(COutPoint&& outpoint, Coin&& coin) {
     cachedCoinsUsage += coin.DynamicMemoryUsage();
-    auto [it, inserted] = cacheCoins.emplace(
-        std::piecewise_construct,
```
Was this verbose code that is replaced here a workaround for lack of `try_emplace`?

l0rinc replied:
It seems this [was introduced in 2017](https://github.com/bitcoin/bitcoin/commit/50830796889ecaa458871f1db878c255dd2554cb#diff-f0ed73d62dae6ca28ebd3045e5fc0d5d02eaaacadb4c2a292985a3fbd7e1c77cR100), when the project was still on [C++11](https://github.com/bitcoin/bitcoin/blob/342b9bc3907edf8eae64440397a32833ed44fae4/configure.ac#L58), but `try_emplace` was only introduced in [C++17](https://en.cppreference.com/w/cpp/container/map/try_emplace).

When this was split out to [`EmplaceCoinInternalDANGER`](https://github.com/bitcoin/bitcoin/commit/f6e2da5fb7c6406c37612c838c998078ea8d2252#diff-095ce1081a930998a10b37358fae5499ac47f8cb6f25f5df5d88e920a54e0341R275), the old impl was copied (on Aug 25, 2020), but C++17 support was only enabled in [Nov 18, 2020](https://github.com/bitcoin/bitcoin/commit/faaee810e62d796d66bfb2bc4e53c8b4c8104d13#diff-3f2b8984fdf0aea4a02d35b54047634465410c1526da78c0b4f888248337d135).

Note that this was already partially upgraded for [FetchCoin](https://github.com/bitcoin/bitcoin/pull/30326/files).

ryanofsky commented on src/test/coins_tests.cpp:
```diff
@@ -764,54 +771,40 @@ static void CheckAddCoinBase(CAmount base_value, CAmount modify_value, const Coi
-    CheckAddCoin(VALUE2, VALUE3, VALUE3, DIRTY      , DIRTY      , true );
-    CheckAddCoin(VALUE2, VALUE3, FAIL  , DIRTY|FRESH, NO_ENTRY   , false);
-    CheckAddCoin(VALUE2, VALUE3, VALUE3, DIRTY|FRESH, DIRTY|FRESH, true );
+    for (auto base_value : {ABSENT, SPENT, VALUE1}) {
```
In commit "test: Group values and states in tests into CoinEntry wrappers" (553db5dec1334ed5e8e07e34306f16de9afc22bb)

Note: because this for loop is moving from `CheckAddCoin` function to the top-level test code, this commit changes order of the tests. But tests that are performed should still be the same. It would be good to note in commit message that this commit does affect test order of this test but doesn't change its coverage.

ryanofsky commented on src/test/coins_tests.cpp:
```diff
@@ -822,81 +815,83 @@ void CheckWriteCoins(const CoinEntry& parent, const CoinEntry& child, const Coin
-            for (const char parent_flags : parent_value == ABSENT ? ABSENT_FLAGS : FLAGS)
-                for (const char child_flags : child_value == ABSENT ? ABSENT_FLAGS : CLEAN_FLAGS)
-                    CheckWriteCoins(parent_value, child_value, parent_value, parent_flags, child_flags, parent_flags);
+    for (const CoinEntry parent : {MISSING,
```
In commit "test: Group values and states in tests into CoinEntry wrappers" (553db5dec1334ed5e8e07e34306f16de9afc22bb)

Note: Again in this part of the test, the order cases are run in seems to be changing, but the same set of cases is run. Would be helpful to mention this in commit message.

l0rinc replied:
Done in previous step

ryanofsky reviewed (approved):
Code review ACK 3b5b2457f713014fd5073da34b76a5a8cd796e4a. Nice change that should make the API more safe and also makes tests more complete and easier to maintain.

l0rinc committed:
coins, refactor: Make AddFlags, SetDirty, SetFresh static

This makes the `Assume(&self.second == this)` check redundant

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

l0rinc committed:
coins, refactor: Assume state after SetClean in AddFlags to prevent dangling pointers

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

l0rinc commented:
Thanks a lot @ryanofsky for the review. Applied almost all of your insightful changes, adding two new commits (adding you as coauthor).
You can view the diff with [`Compare`](https://github.com/bitcoin/bitcoin/compare/3b5b2457f713014fd5073da34b76a5a8cd796e4a..aa2f3139529c054b011a0f75ff314e6d63f0d977) or via `git range-diff 8e1381a..3b5b2457 a6921049..aa2f3139`

Added the following commits based on feedback:
* coins, refactor: Assume state after SetClean in AddFlags to prevent dangling pointers
* coins, refactor: Make AddFlags, SetDirty, SetFresh static

Updated commit messages to warn about test order changes and method move reasons.
Separated method move from method changes and other minor moves between commits for clarity.
Asserted both fresh and dirty in tests where flags were changed originally.

l0rinc commented on src/test/coins_tests.cpp:
```diff
@@ -661,47 +676,43 @@ static void CheckAccessCoin(CAmount base_value, const CoinEntry& cache_coin, con
-    CheckAccessCoin(VALUE1, VALUE2, VALUE2, FRESH      , FRESH      );
-    CheckAccessCoin(VALUE1, VALUE2, VALUE2, DIRTY      , DIRTY      );
-    CheckAccessCoin(VALUE1, VALUE2, VALUE2, DIRTY|FRESH, DIRTY|FRESH);
+    CheckAccessCoin(ABSENT, MISSING,            MISSING           );
```
Thank you for checking! Do you need any change here from me?

ryanofsky replied:
Thanks, no this is just a note to help myself and other reviewers.

l0rinc commented on src/test/coins_tests.cpp:
```diff
@@ -558,16 +560,32 @@ BOOST_AUTO_TEST_CASE(ccoins_serialization)
+    friend std::ostream& operator<<(std::ostream& os, const CoinEntry& e) { return os << e.value << e.flags; }
+
+    constexpr bool IsDirty() const { return flags & DIRTY; }
+    constexpr bool IsFresh() const { return flags & FRESH; }
```
Done

l0rinc commented on src/test/coinscachepair_tests.cpp:
```diff
@@ -63,7 +63,7 @@ BOOST_AUTO_TEST_CASE(linked_list_iteration)
     // Delete the nodes from the list to make sure there are no dangling pointers
     for (auto it{nodes.begin()}; it != nodes.end(); it = nodes.erase(it)) {
-        BOOST_CHECK_EQUAL(it->second.GetFlags(), 0);
+        BOOST_CHECK(!it->second.IsDirty() && !it->second.IsFresh());
```
Yeah, but I don't like the format of that. If this fails, it's easy to debug.

l0rinc commented on src/test/coinscachepair_tests.cpp:
```diff
@@ -21,7 +21,7 @@ std::list<CoinsCachePair> CreatePairs(CoinsCachePair& sentinel)
         node->second.SetDirty(*node, sentinel);
 
-        BOOST_CHECK_EQUAL(node->second.GetFlags(), CCoinsCacheEntry::DIRTY);
+        BOOST_CHECK(node->second.IsDirty());
```
Hmm, so always check both? Ok, I don't mind, done

l0rinc commented on src/coins.h:
```diff
@@ -156,24 +172,18 @@ struct CCoinsCacheEntry
+    inline void SetClean() noexcept
     {
         if (!m_flags) return;
         m_next->second.m_prev = m_prev;
```
Good idea, did it in a separate commit, let's see what the CI and others think.

l0rinc commented on src/coins.h:
```diff
@@ -156,24 +172,18 @@ struct CCoinsCacheEntry
-    //! this entry in the CCoinsCache map and a reference to the sentinel of the
-    //! flagged pair linked list.
-    inline void AddFlags(uint8_t flags, CoinsCachePair& self, CoinsCachePair& sentinel) noexcept
+    inline void SetDirty(CoinsCachePair& self, CoinsCachePair& sentinel) noexcept
```
Or maybe even go one step further and make `SetDirty`/`SetFresh` static as well - thanks for the hint, pushed in a separate commit with your coauthorship.

andrewtoth replied:
This is a great idea!

l0rinc commented on src/coins.h:
```diff
@@ -156,24 +172,18 @@ struct CCoinsCacheEntry
-    //! Adding a flag also requires a self reference to the pair that contains
-    //! this entry in the CCoinsCache map and a reference to the sentinel of the
-    //! flagged pair linked list.
-    inline void AddFlags(uint8_t flags, CoinsCachePair& self, CoinsCachePair& sentinel) noexcept
```
added refactor to the first two commits, thanks

l0rinc commented on src/test/coins_tests.cpp:
```diff
@@ -747,25 +754,22 @@ BOOST_AUTO_TEST_CASE(ccoins_spend)
-    CAmount result_value;
-    char result_flags;
     try {
+        SingleEntryCacheTest test(base_value, cache_coin);
```
I was hesitant at first (mostly because this will be removed in `test: Validate error messages on fail`), but we want to always fail when these do, so you're right, thanks!

l0rinc commented on src/coins.h:
```diff
@@ -128,6 +128,22 @@ struct CCoinsCacheEntry
+    //! Adding a flag also requires a self reference to the pair that contains
+    //! this entry in the CCoinsCache map and a reference to the sentinel of the
+    //! flagged pair linked list.
+    inline void AddFlags(uint8_t flags, CoinsCachePair& self, CoinsCachePair& sentinel) noexcept
```
Good idea, done!

ryanofsky commented on src/coins.h:
```diff
@@ -166,6 +166,7 @@ struct CCoinsCacheEntry
     {
         Assume(flags & (DIRTY | FRESH));
         if (!self.second.m_flags) {
+            Assume(self.second.m_prev == nullptr && self.second.m_next == nullptr);
```
In commit "coins, refactor: Assume state after SetClean in AddFlags to prevent dangling pointers" (2d691b50d3729445a60898981b78c4239f2f0dd7)

I think this could also check the converse, that pointers are valid if flags were set. It could simplify check that pointers are non-null if flags are set, or go further and check prev->next == self, and next->prev == self

l0rinc replied:
> check that pointers are non-null if flags are set, or go further and check prev->next == self, and next->prev == self

the latter seems redundant, but the first is simple to check - done

ryanofsky commented on src/test/coins_tests.cpp:
```diff
@@ -558,16 +560,29 @@ BOOST_AUTO_TEST_CASE(ccoins_serialization)
+const static char FRESH = CCoinsCacheEntry::FRESH;
+const static char NO_ENTRY = -1;
+
+struct CoinEntry {
```
In commit "test: Migrate GetCoinsMapEntry to return MaybeCoin" (a902af6fb195fae97a89876b88d5034162ab8366)

May want to add "refactor" to commit subject or otherwise mention that this commit is not adding or removing any test coverage. Otherwise it's hard to know looking at the commit what it's intended to do.

Same comment also applies to commit "test: Remove remaining unbounded flags from coins_tests" (e4db49949c55470950a206172ee53f6baacf3767) and commit "test: Compact ccoins_access and ccoins_spend" (aa2f3139529c054b011a0f75ff314e6d63f0d977)

l0rinc replied:
Done

ryanofsky commented on src/test/coins_tests.cpp:
```diff
@@ -558,16 +560,29 @@ BOOST_AUTO_TEST_CASE(ccoins_serialization)
+    CoinEntry(CAmount v, char s) : value(v), flags(s) {}
+
+    bool operator==(const CoinEntry& o) const = default;
+    friend std::ostream& operator<<(std::ostream& os, const CoinEntry& e) { return os << e.value << e.flags; }
```
In commit "test: Migrate GetCoinsMapEntry to return MaybeCoin" (a902af6fb195fae97a89876b88d5034162ab8366)

Should this include a space between that value and flags? Otherwise if test fails it looks like amount and flags will run together and might not be possible to distinguish.

l0rinc replied:
Done

ryanofsky reviewed (approved):
Code review ACK aa2f3139529c054b011a0f75ff314e6d63f0d977. Thanks for the updates! Left a few more comments that are not important and you can feel free to ignore.

andrewtoth commented on src/coins.h:
```diff
@@ -162,26 +162,19 @@ struct CCoinsCacheEntry
     //! Adding a flag also requires a self reference to the pair that contains
     //! this entry in the CCoinsCache map and a reference to the sentinel of the
```
nit: this comment is no longer accurate after d5e3bbf440aa948df8159999fd4eb5275c354b93

We should just remove the part about a self reference, and only mention the sentinel.

andrewtoth commented on src/coins.h:
```diff
@@ -162,26 +162,19 @@ struct CCoinsCacheEntry
     //! this entry in the CCoinsCache map and a reference to the sentinel of the
     //! flagged pair linked list.
-    inline void AddFlags(uint8_t flags, CoinsCachePair& self, CoinsCachePair& sentinel) noexcept
+    static inline void AddFlags(uint8_t flags, CoinsCachePair& self, CoinsCachePair& sentinel) noexcept
```
I think `self` is no longer a good name for this parameter if we make this static. It should just be `pair`.

Same for `SetDirty` and `SetFresh`.

l0rinc replied:
Done

andrewtoth commented on src/test/coinscachepair_tests.cpp:
```diff
@@ -139,56 +139,55 @@ BOOST_AUTO_TEST_CASE(linked_list_random_deletion)
     BOOST_CHECK_EQUAL(n2.second.Prev(), &n1);
 
-    // Check that we can clear flags then re-add them
+    // Check that we can clear state then re-set them
```
nit: `// Check that we can clear state then re-set it`

l0rinc replied:
Done

andrewtoth reviewed (approved):
re-crACK aa2f3139529c054b011a0f75ff314e6d63f0d977

It might be worth updating the `AddFlags` comment now, but otherwise non-blocking nits.

l0rinc commented on src/coins.h:
```diff
@@ -156,24 +156,33 @@ struct CCoinsCacheEntry
+    {
+        AddFlags(DIRTY, self, sentinel);
+    }
+    inline void SetFresh(CoinsCachePair& self, CoinsCachePair& sentinel) noexcept
```
Wouldn't I need to make `AddFlags` static as well in that case?
Which would trigger `this` removal and rewrite of the whole `AddFlags` method which is done in the next commit. Or you mean squash the first two commits?

l0rinc commented:
Thanks @ryanofsky & @andrewtoth, applied most nits, please re-review: `git range-diff a6921049..aa2f3139 a6921049..9edbe5a4` or https://github.com/bitcoin/bitcoin/compare/3b5b2457f713014fd5073da34b76a5a8cd796e4a..9edbe5a48fe9a2c0d364882fcb7d44ae6aac6d24

andrewtoth commented:
re-crACK 9edbe5a48fe9a2c0d364882fcb7d44ae6aac6d24

andrewtoth commented on src/coins.h:
```diff
@@ -197,11 +201,11 @@ struct CCoinsCacheEntry
 
     //! Only use this for initializing the linked list sentinel
-    inline void SelfRef(CoinsCachePair& self) noexcept
+    inline void SelfRef(CoinsCachePair& pair) noexcept
```
minor nit: I don't think this one needed to be renamed to `pair`. `self` still makes sense here.

l0rinc replied:
yeah, but it's more consistent this way (and method name already mentions self)

ryanofsky reviewed (approved):
Code review ACK 9edbe5a48fe9a2c0d364882fcb7d44ae6aac6d24. Only small changes since last review like adding an Assume, changing commit messages, renaming self to pair. Thanks for the updates!

I left two suggestions which are not important and don't affect the final code, just meant to make review a little easier for the next reviewer. Feel free to ignore them or save for a later update.

hodlinator commented on src/test/coins_tests.cpp:
```diff
@@ -585,314 +624,245 @@ static void SetCoinsValue(CAmount value, Coin& coin)
     {
-        WriteCoinsViewEntry(base, base_value, base_value == ABSENT ? NO_ENTRY : DIRTY);
-        cache.usage() += InsertCoinsMapEntry(cache.map(), cache.sentinel(), cache_value, cache_flags);
+        auto base_cache_coin{base_value == ABSENT ? MISSING : CoinEntry{base_value, CoinEntry::State::DIRTY}};
```
nit: Going from `DIRTY` to `CoinEntry::State::DIRTY` is a bit noisy. I'd prefer making `CoinEntry::State` a non-`class` `enum`.
```suggestion
        auto base_cache_coin{base_value == ABSENT ? MISSING : CoinEntry{base_value, CoinEntry::DIRTY}};
```

l0rinc replied:
enum class is safer than enum https://agrawalsuneet.github.io/blogs/enum-vs-enum-class-in-c++ and I didn't like the result of `using enum State`

hodlinator reviewed (commented):
Code reviewed 9edbe5a48fe9a2c0d364882fcb7d44ae6aac6d24

Good that `AddFlags()` was made `static` so one doesn't have both `this` and `self` any longer! 😅

Surprised that `CCoinsCacheEntry::m_next` and `m_prev` were not nulled out before this PR as is done in `SetClean()` in 9d61b956d91153d25d06047de03133bc2e23b2a6, but seems reasonable. Only caller of `SetClean()` outside of tests is `CoinsViewCacheCursor::NextAndMaybeErase()`. That's the only commit where I feel that more domain expertise would help calm me down. **If anyone has any assuring context to drop on me, please do.**

It would be good to make `AddFlags()` `private` in the commit after introducing `SetDirty()` & `SetFresh()` to further clarify that nothing exterior is using it any longer. Also think it may be good to introduce `SetDirty()` & `SetFresh()` as `static` from the start to reduce churn. Doing so makes `AddFlags` a bit harder to review though.

Branch with my suggested changes:
https://github.com/bitcoin/bitcoin/compare/master...hodlinator:bitcoin:pr/30906_suggestions (commit making `AddFlags()` & co `static` is merged into first commit).
[Comparison at this PR.](https://github.com/bitcoin/bitcoin/compare/9edbe5a48fe9a2c0d364882fcb7d44ae6aac6d24..840924d5b6d88d6dc841517fade8a78745322e52)

l0rinc commented on src/test/coins_tests.cpp:
```diff
@@ -585,314 +624,245 @@ static void SetCoinsValue(CAmount value, Coin& coin)
+        for (const MaybeCoin& child : {MISSING,
+                                       SPENT_CLEAN, SPENT_FRESH,
+                                       VALUE2_CLEAN, VALUE2_FRESH}) {
+            auto expected{CoinOrError{parent}}; // TODO add failure cases
```
yes, see the mentioned derisked PR: https://github.com/bitcoin/bitcoin/pull/30673/files#diff-1ef3b6a1936b50f3d5ec4a1786d9e2d63d1a3e1815b103e67f20601995f355b4R152-R154

hodlinator replied:
Don't get it, what does *src/test/fuzz/coinscache_sim.cpp* have to do with this? Something with clearing coins?
By "failure", do you mean the error half of `CoinOrError`?

l0rinc commented on src/test/coins_tests.cpp:
```diff
@@ -585,314 +624,245 @@ static void SetCoinsValue(CAmount value, Coin& coin)
-        }
-        flags = it->second.GetFlags();
-        assert(flags != NO_ENTRY);
+    if (auto it{map.find(outp)}; it != map.end()) {
```
If I touch again, I'll fix the curlies :)

l0rinc replied:
Done

l0rinc commented on src/test/coinscachepair_tests.cpp:
```diff
@@ -139,77 +139,55 @@ BOOST_AUTO_TEST_CASE(linked_list_random_deletion)
 
-    // Adding 0 still has no effect
-    n1.second.AddFlags(0, n1, sentinel);
+    n1.second.SetClean();
```
Done

l0rinc commented on src/test/coins_tests.cpp:
```diff
@@ -585,314 +624,245 @@ static void SetCoinsValue(CAmount value, Coin& coin)
 }
 
-static void CheckAddCoinBase(CAmount base_value, CAmount cache_value, CAmount modify_value, CAmount expected_value, char cache_flags, char expected_flags, bool coinbase)
+static void CheckAddCoin(CAmount base_value, CAmount modify_value, const MaybeCoin& cache_coin, const CoinOrError expected, bool coinbase)
```
I must have had a reason for changing it, but it does make sense to have the cache value before the modify value in the row -> done

l0rinc commented on src/test/coins_tests.cpp:
```diff
@@ -585,314 +624,245 @@ static void SetCoinsValue(CAmount value, Coin& coin)
-        test.cache.AddCoin(OUTPOINT, Coin(std::move(output), 1, coinbase), coinbase);
+    SingleEntryCacheTest test(base_value, cache_coin);
+    bool possible_overwrite{coinbase};
+    auto add_coin{[&]() { test.cache.AddCoin(OUTPOINT, Coin(CTxOut(modify_value, CScript()), 1, coinbase), possible_overwrite); }};
```
Done

l0rinc commented on src/test/coins_tests.cpp:
```diff
@@ -558,20 +560,57 @@ BOOST_AUTO_TEST_CASE(ccoins_serialization)
+using CoinOrError = std::variant<MaybeCoin, std::string>;
+
+constexpr MaybeCoin MISSING           {std::nullopt};
+constexpr MaybeCoin SPENT_DIRTY       {CoinEntry{SPENT,  CoinEntry::State::DIRTY}};
```
Done

l0rinc commented on src/test/coins_tests.cpp:
```diff
@@ -558,20 +560,57 @@ BOOST_AUTO_TEST_CASE(ccoins_serialization)
+    const CAmount value;
+    const State state;
+
+    constexpr CoinEntry(CAmount v, State s) : value(v), state(s) {}
```
Done

l0rinc commented on src/test/coins_tests.cpp:
```diff
@@ -15,6 +15,8 @@
 #include <util/strencodings.h>
 
 #include <map>
+#include <string>
+#include <variant>
```
Done

hodlinator replied:
`<string>` is still needlessly added in c0b4b2c1eef95c0b626573a9a2e217f4a541a880.

l0rinc replied:
```bash
grep "std::string" bitcoin/src/test/coins_tests.cpp | wc -l
       3
```

l0rinc commented on src/test/fuzz/coins_view.cpp:
```diff
@@ -140,7 +141,8 @@ FUZZ_TARGET(coins_view, .init = initialize_coins_view)
                     auto it{coins_map.emplace(random_out_point, std::move(coins_cache_entry)).first};
-                    it->second.AddFlags(flags, *it, sentinel);
+                    if (dirty) it->second.SetDirty(*it, sentinel);
+                    if (fresh) it->second.SetFresh(*it, sentinel);
```
Done, thanks

laanwj reviewed (approved):
Code review ACK 50cce20013c97e1257cb9f4c9319701a09b0a196

i have checked that behavior stays the same with regard to adding DIRTY and FRESH flags, i did not check all the test changes in detail but overall LGTM.

hodlinator commented on src/test/coins_tests.cpp:
```diff
@@ -559,21 +561,58 @@ BOOST_AUTO_TEST_CASE(ccoins_serialization)
+    const CAmount value;
+    const State state;
+
+    constexpr CoinEntry(const CAmount v, const State s) : value{v}, state{s} {}
```
Using `const` for value type parameters (`CAmount is a typedef of `int64_t`), when a function doesn't contain any logic is noisy to me and not used in the codebase at large.
```suggestion
    constexpr CoinEntry(CAmount v, State s) : value{v}, state{s} {}
```

It is useful in cases like this where the function is actually doing something:
https://github.com/bitcoin/bitcoin/blob/ebe4cac38bf6c510b00b8e080acab079c54016d6/src/uint256.cpp#L20-L44

Same goes for `ToState`, `SetCoinsValue`, `SingleEntryCacheTest`, `CheckAccessCoin` etc.

Otherwise you might as well add `const` to `AddFlags(uint8_t flags, ...` etc?

l0rinc replied:
I prefer `const` here, my IDE was also asking for them, I don't mind adding them here.
Being `noisy` when `a function doesn't contain any logic` doesn't sound like a blocker to me :)

hodlinator replied:
This issue is a non-nit for me as a general pattern as it will affect diffs & reviews going forward.

We've got a ratio of 5/212 or roughly 1:40 for this codebase. I'm surprised that the IDE is encouraging that.

What is you rationale for not applying it to `AddFlags`? Did the specific version of your IDE not tell you to do so? Does your IDE have a documented rationale?

If there is a clear rationale, ideally with a statistical sample of % bugs discovered from changing to `const` value type parameters, I'm open to that. But the 1:40 ratio weighs heavily.

l0rinc replied:
We can unify these in a separate PR, I don't mind, I added these since the linters were complaining - and I was fine with either option since it doesn't affect the body of the function signal-to-noise-ratio-wise.

hodlinator replied:
9edbe5a48fe9a2c0d364882fcb7d44ae6aac6d24 before the latest push didn't have these `const` params. I think it was a mistake to add them in the latest push and would prefer them rolled back.

Another alternative is to discuss my samples of ratios of `const` value type params or why you didn't apply it to `AddFlags`.

Sorry to be annoying about this but for some reason it really annoys me. :)
Will take a timeout and see if I can let go of this, if you still insist.

ryanofsky reviewed (approved):
Code review ACK 50cce20013c97e1257cb9f4c9319701a09b0a196. Looks good! Thanks for the followups.

Since last review, coins.h commits were reorganized to minimize diffs, but only overall change was to drop redundant `inline` keywords. In `coins_tests.cpp` a lot of smaller changes were made like adding `const` to amount parameters (which is not great but ok), using more brace initialization, changing CheckAddCoin parameter order, and fixing up some comments.

davidgumberg commented:
post-merge ACK https://github.com/bitcoin/bitcoin/commit/50cce20013c97e1257cb9f4c9319701a09b0a196

This refactor reduces the risk of  `CCoinsCacheEntry`'s interfaces, and making `CCoinsCacheEntry::AddFlags()` static is a big improvement for readability. 

I did not thoroughly review the refactors to unit tests, but they seem correct to me.

More broadly, I think the changes made here to `SetClean()` are good, but `SetClean()` being public (only used externally by `NextAndMaybeErase()`) seems dangerous to me. The invariant that the code currently respects is that a coin's state has been written to the parent/backing view before `NextAndMaybeErase()`->`SetClean()`, so removing DIRTY and FRESH flags are fine, but is it risky to have an interface that doesn't enforce that?

hodlinator commented:
I stand by my refusal to fully ACK based off the unrequested addition of `const` parameter noise in the last push, but aside from that the PR is good.

Follow-up idea: Noticed `CCoinsCacheEntry::Flags` can be made private now that `CoinEntry` in the test has its own `enum`.

l0rinc commented:
> the changes made here to SetClean() are good, but SetClean() being public (only used externally by NextAndMaybeErase()) seems dangerous to me. 

Can you provide a PR for it? We need to clean up this area in tiny steps, if you can objectively make it better, we'll review it.

> Follow-up idea: Noticed CCoinsCacheEntry::Flags can be made private now that CoinEntry in the test has its own enum.

Seems you had strong preferences in other areas as well, follow-up PRs are welcome.

hodlinator commented:
> > Follow-up idea: Noticed CCoinsCacheEntry::Flags can be made private now that CoinEntry in the test has its own enum.
> 
> ..., follow-up PRs are welcome.

#31496
